source/fundamentals/logging.txt

```diff
@@ -0,0 +1,273 @@
+         MongoDatabase database = mongoClient.getDatabase(<my database>);
+         MongoCollection<Document> collection = database.getCollection(<my collection>);
+         collection.find().first();
+
+      You should see output resembling the following:
+
+      .. code-block:: none
+
+         ...
+         1156 [main] DEBUG o.m.driver.protocol.command - Sending command '{"find": "<your collection>", "filter": {}, "limit": 1, "singleBatch": true, "$db": "<your db>", "lsid": {"id": {"$binary": {"base64": "<server session>", "subType": "04"}}}}' with request id 13 to database <your db> on connection <your connection> to server <your server>
+         1173 [main] DEBUG o.m.driver.protocol.command - Execution of command with request id 13 completed successfully in 30.01 ms on connection <your connection> to server <your server>
+   
+      For more information on Logback, see the
+      `Logback manual <http://logback.qos.ch/manual/>`__.
+   
+   .. tab:: Log4j2
+      :tabid: Log4j2-binding
+   
+      If you add these dependencies to your ``pom.xml`` file, the MongoDB Java
+      driver uses the Log4j2 framework to output your logs.
+
+      .. code-block:: xml      
+      
+         <!-- SLF4J API -->
+         <dependency>
+           <groupId>org.slf4j</groupId>
+           <artifactId>slf4j-api</artifactId>
+           <version>1.7.30</version>
+         </dependency>
+         <!-- SLF4J-Log4j2 binding -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-slf4j-impl</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+         <!-- Log4j2 core functionality -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-core</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+
+      Once you have installed the above dependencies, try logging an error.
+
+      .. code-block:: java
+
+         import org.slf4j.Logger;
+         import org.slf4j.LoggerFactory;
+
+         ...
+
+         Logger logger = LoggerFactory.getLogger(App.class);
+         logger.error("Logging an Error");
+
+      You should see output resembling the following:
+
+      .. code-block:: none
+
+         12:35:00.438 [main] ERROR <my package path> - Logging an Error
+
+      For more information on Log4j2, see the
+      `Log4j2 manual <https://logging.apache.org/log4j/2.x/manual/>`__.
+
+Configure Your Logger
+---------------------
+
+To configure your logger, you must configure the logging framework bound to
+SLF4J.
+
+The example below shows how to configure your logger's log level.
+
+Log levels are the hierarchy of messages output by a logger based on urgency.
+Messages related to errors are the most urgent, and messages related to
+stack traces and debugging are the least urgent. A logger's log level
+specifies a lower bound for how urgent a message must be for the logger to
+output it.
```

Comment:
I'm not sure "urgent" is correct? My thinking is that urgency is assigned to labels by the developer, "error" and "warning" are just labels unless they are being semantically parsed.

I'm unable to fully determine the subject of this paragraph.  It also feels like fundamental information that should be presented at the beginning of the section.

Reply:
Thank you for the note. I disagree that log level labels do not have meaning. The `Log4J2` default log levels are [placed in a hierarchy by corresponding integer values ](https://logging.apache.org/log4j/2.x/manual/customloglevels.html) giving them a very real sort order. The log4j2 source code [documents standard log levels by severity ](https://github.com/apache/logging-log4j2/blob/79d668afff389b4d9b8c1bd9af26269e28667597/log4j-1.2-api/src/main/java/org/apache/log4j/Level.java#L54). I think severity of message corresponds to urgency of responding to log. Based on this, I believe urgent is the correct term.

If you strongly disagree with this I can remove it, but I believe that log levels have semantic meaning in code.